Add post interp callback to dump metric data in Bondi-Sachs form

knelli2 opened the pull request:
## Proposed changes

This adds a new PostInterpolationCallback that will take metric data on concentric extraction spheres, convert it to the Bondi-Sachs form, and dump it to a separate file for each extraction radius. The format of the output data and the filenames are chosen so that it can be read in with the CCE executable.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments
~Depends on and includes #4460.~

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Add DumpBondiSachsOnWorldtube post interpolation callback

knelli2 committed:
Add tag for dumping bondi sachs data

wthrowe commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -53,6 +53,15 @@ struct CceEvolutionPrefix {
+struct BondiSachsOutputFilePrefix {
+  using type = std::string;
+  static constexpr Options::String help{
+      "Filename prefix for dumbing Bondi-Sachs data on worltube radii. Files "
```
"dumping"

wthrowe commented on src/Evolution/Systems/Cce/Callbacks/DumpBondiSachsOnWorldtube.hpp:
```diff
@@ -0,0 +1,301 @@
+        Parallel::get<Tags::Sphere<InterpolationTargetTag>>(cache);
+    const auto& filename_prefix = Parallel::get<Cce::Tags::FilePrefix>(cache);
+
+    ASSERT(sphere.angular_ordering == intrp::AngularOrdering::Cce,
```
ERROR, as this is controlled from the input file.

wthrowe commented on src/Evolution/Systems/Cce/Callbacks/DumpBondiSachsOnWorldtube.hpp:
```diff
@@ -0,0 +1,301 @@
+ * - `gr::Tags::SpacetimeMetric`
+ * - `GeneralizedHarmonic::Tags::Pi`
+ * - `GeneralizedHarmonic::Tags::Phi`
+ * - `::Tags::deriv<GeneralizedHarmonic::Tags::Phi>`
```
I don't see any uses of `deriv<Phi>`.  Why does it have to be there?

wthrowe commented on src/Evolution/Systems/Cce/Callbacks/DumpBondiSachsOnWorldtube.hpp:
```diff
@@ -0,0 +1,301 @@
+    const auto& all_phi =
+        get<GeneralizedHarmonic::Tags::Phi<3, ::Frame::Inertial>>(all_gh_vars);
+
+    Variables<gh_source_vars_for_cce> tmp_buffer;
```
You never allocate any memory in this buffer.  It's not saving you anything, and I don't think changing the `data()` value in a `Variables`-backed `Tensor` is a supported operation.

knelli2 replied:
Well I need a `Variables` of g, pi, phi that only has one sphere's worth of points in order to pass to `create_bondi_data` for each extraction radius. Any suggestions on how to do that then without a copy? I already have all the data in `all_gh_vars` for all extraction spheres, Just need to split it up a bit.

wthrowe replied:
You're not passing this to anything, and it has zero points.

Avoiding the copies with non-owning DataVectors is good.  I just don't understand why you want an empty temporary buffer.

knelli2 replied:
Oh I see what you mean. I used a `Variables` just to easily get un-allocated tensors for g, pi, and phi so I could set their data refs when I loop over the radii. I guess I could have explicitly wrote out the unallocated tensors, but that was more lines so I didn't.

wthrowe replied:
It's three lines instead of two.

knelli2 replied:
Ok I changed it.

wthrowe commented on src/Evolution/Systems/Cce/Callbacks/DumpBondiSachsOnWorldtube.hpp:
```diff
@@ -0,0 +1,301 @@
+      // Bondi data
+      Variables<cce_boundary_tags> bondi_boundary_data{
+          num_points_single_sphere};
+      ComplexModalVector goldberg_mode_buffer{square(l_max + 1)};
```
These two *are* allocations inside the loop, so moving them outside would help.

wthrowe commented on src/Evolution/Systems/Cce/Callbacks/DumpBondiSachsOnWorldtube.hpp:
```diff
@@ -0,0 +1,301 @@
+                                      pi, spacetime_metric, radius, l_max);
+
+      const std::vector<std::string> all_legend = build_legend(l_max, false);
+      const std::vector<std::string> real_legend = build_legend(l_max, true);
```
These two lines may cause the majority of the allocations in this function.  Move them out of the loop.

wthrowe commented on src/Evolution/Systems/Cce/Callbacks/DumpBondiSachsOnWorldtube.hpp:
```diff
@@ -0,0 +1,301 @@
+            // because we don't print negative m or the imaginary part of m=0
+            // for real quantities.
+            for (size_t ell = 0; ell <= l_max; ell++) {
+              for (int m = is_real ? 0 : -ell; m <= static_cast<int>(ell);
```
Cast the first `ell` too.

knelli2 committed:
Fix variables in TryToInterpolate

Previously if the compute_vars_to_interpolate alias wasn't in the target tag,
then TryToInterpolate would just use the interpolator_source_vars. However,
if the compute_vars_to_interpolate isn't in the target tag, but the
vars_to_interpolate_to_target alias is specified as a subset of the
interpolator_source_vars then the wrong if constexpr branch is taken. This
fixes it to use the subset of vars_to_interpolate_to_target

wthrowe reviewed (dismissed):
Looks good.  Squash.

nilsdeppe reviewed (approved):
I don't love that we need a new file for each extraction radius, but I assume that's a current requirement of CCE, right? I'm fine leave as-is for now, but we should probably combine them into the reductions file at some point (or maybe one CCE file with subfiles)

nilsdeppe commented:
CI failure is timeout in binary compact object test.

knelli2 commented:
@nilsdeppe Yes the reason I did it like this is because currently CCE is set up to read in metric data which is output from SpEC in a specific format (different files for each radius, named a certain way). I agree eventually it'd be nice to have CCE read in data from SpECTRE in our own format, but that's a whole change to CCE which is why I didn't bother doing it here. Just wanted to get waveforms working quickly.

nilsdeppe commented:
Okay, thanks for clarifying! I completely agree, getting it into one file is a much bigger change and very low priority :) Thanks!
